lib/active_admin/resource_controller/decorators.rb

```diff
@@ -48,18 +48,45 @@ def collection_decorator
         end
       end
 
+      class CollectionDecorator
+        # @return the collection being decorated.
+        attr_reader :object
+
+        # @return [Class] the decorator class used to decorate each item, as set by {#initialize}.
+        attr_reader :decorator_class
+
+        array_methods = Array.instance_methods - Object.instance_methods
+        delegate :==, :as_json, *array_methods, to: :decorated_collection
+
+        def initialize(object, with:)
+          @object = object
+          @decorator_class = with
+        end
+
+        def self.name
+          'ActiveAdmin::CollectionDecorator'
+        end
+
+        class << self
+          alias_method :decorate, :new
+        end
+
+        def decorated_collection
+          @decorated_collection ||= object.map { |item| decorator_class.new(item) }
+        end
+      end
+
       class Wrapper
         @cache = {}
 
         def self.wrap(decorator)
-          collection_decorator = find_draper_collection_decorator(decorator)
-
-          if collection_decorator
-            name = "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
-            @cache[name] ||= wrap! collection_decorator, name
-          else
-            decorator
-          end
+          collection_decorator = find_collection_decorator(decorator)
+          name = if collection_decorator <= CollectionDecorator
+                   "#{collection_decorator.name} of #{decorator}"
+                 else
+                   "#{collection_decorator.name} of #{decorator} + ActiveAdmin"
```

Comment:
Are the two branches here needed? I guess the `+ ActiveAdmin` part makes sense either way since we're adding an extra layer on top of the raw collection decorator.